cmake: support find_package(XNNPACK)

ConnorBaker opened the pull request:
Adds a CMake configuration file to make it possible for CMake-based projects to use `find_package` to consume this library.

Part of the work being done for https://github.com/NixOS/nixpkgs/pull/239291.

ConnorBaker commented:
@Maratyszcza @alankelly would either of you have a chance to review this?

Maratyszcza commented on cmake/XNNPACKConfig.cmake.in:
```diff
@@ -0,0 +1,5 @@
+@PACKAGE_INIT@
```
Use uppercase for CMake commands

ConnorBaker replied:
Resolved in latest

Maratyszcza commented on CMakeLists.txt:
```diff
@@ -825,7 +826,9 @@ IF(XNNPACK_BUILD_LIBRARY)
   TARGET_INCLUDE_DIRECTORIES(mutex PRIVATE include src)
-  TARGET_INCLUDE_DIRECTORIES(post-operation PUBLIC include src)
+  TARGET_INCLUDE_DIRECTORIES(post-operation PUBLIC
+    $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>
```
Why do you need to specify `$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>` instead of just `include`? Why do you do it only for some targets (`XNNPACK`, and `post-operation`)?

ConnorBaker replied:
If we use

```cmake
TARGET_INCLUDE_DIRECTORIES(XNNPACK PUBLIC include)
```

with the CMake configuration which allows discovery of XNNPACK after installation, then trying to build will give the following error:

```console
CMake Error in CMakeLists.txt:
  Target "XNNPACK" INTERFACE_INCLUDE_DIRECTORIES property contains path:
     "/build/source/include"
  which is prefixed in the source directory.
```

CMake interprets `include` as a relative path and converts it to an absolute path based on the current working directory. This path would double as the install path, which is not what we want (that build path wouldn't get a good place for it to look for an install).

Using the `BUILD_INTERFACE` and `INSTALL_INTERFACE` generators allow us a way of saying "when building, use this path, but when installing, use this other path." 

It was wrong of me to include it on `post-operation` -- the public header is coming from `XNNPACK`, so only `XNNPACK` needs to be configured with an understanding of where to find the header at build time, and where to put it during installation.

Maratyszcza commented on cmake/XNNPACKConfig.cmake.in:
```diff
@@ -0,0 +1,5 @@
+@PACKAGE_INIT@
+INCLUDE(CMakeFindDependencyMacro)
+FIND_DEPENDENCY(pthreadpool REQUIRED)
```
What about other dependencies (e.g. `cpuinfo`)?

ConnorBaker replied:
Since `cpuinfo` is linked privately

https://github.com/google/XNNPACK/blob/0961e3449544d147a8486f43a3fdbb4e5d66e162/CMakeLists.txt#L832

we're signaling to CMake that `cpuinfo` is a build-time only dependency -- an implementation detail of the project that consumers of the project don't need to worry about.

Conversely, because `pthreadpool` is linked publicly

https://github.com/google/XNNPACK/blob/0961e3449544d147a8486f43a3fdbb4e5d66e162/CMakeLists.txt#L864

we're signaling to CMake that `pthreadpool` is a run-time dependency and that anyone who wants to use `XNNPACK` should make sure `pthreadpool` is available.

ConnorBaker commented:
Rebased, squashed, and force-pushed!

@Maratyszcza I think I answered your comments, please let me know if you have any more questions.

I apologize for the delay in getting back to you!

xbwee1024 commented:
I very  much agree with this PR, because if i just use "xnnpack.h" and "libXNNPACK.a" after installed into /usr/local , then a lot of "undefined reference... " build error will occur!

ConnorBaker commented on CMakeLists.txt:
```diff
@@ -923,10 +925,22 @@ IF(XNNPACK_BUILD_LIBRARY)
+
+  INCLUDE(CMakePackageConfigHelpers)
+  CONFIGURE_PACKAGE_CONFIG_FILE(
+    ${CMAKE_CURRENT_SOURCE_DIR}/cmake/${CMAKE_PROJECT_NAME}Config.cmake.in
```
Good catch! Fixed in latest.

ConnorBaker commented on CMakeLists.txt:
```diff
@@ -923,10 +925,22 @@ IF(XNNPACK_BUILD_LIBRARY)
+    ${CMAKE_CURRENT_SOURCE_DIR}/cmake/${CMAKE_PROJECT_NAME}Config.cmake.in
+    ${CMAKE_CURRENT_BINARY_DIR}/${CMAKE_PROJECT_NAME}Config.cmake
+    INSTALL_DESTINATION ${CMAKE_INSTALL_LIBDIR}/cmake/${CMAKE_PROJECT_NAME}
+  )
```
Fixed in latest, sorry about that.

ConnorBaker commented on CMakeLists.txt:
```diff
@@ -773,7 +773,9 @@ TARGET_INCLUDE_DIRECTORIES(normalization PRIVATE include src)
 IF(XNNPACK_BUILD_LIBRARY)
-  TARGET_INCLUDE_DIRECTORIES(XNNPACK PUBLIC include)
+  TARGET_INCLUDE_DIRECTORIES(XNNPACK PUBLIC
+    $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>
```
Better safe than sorry -- fixed in latest :)

ConnorBaker commented:
Quoted the path names to avoid weirdness when interpolating strings. (Yay, strings!)

Sorry for the horrible average turn around time!
